### PR TITLE
Analyzer plugins

### PR DESCRIPTION
**Pull request by lrytz**
various fixes, cleanups in namers. new: analyzer plugins. check individual commit messages.

review by @paulp, @retronym

### CODE REVIEW

**Comment by lrytz** [comment]
do we have two kitties now?

**Comment by xeno-by** [comment]
The PR is submitted against 2.10.x. Is this intentional?

**Comment by lrytz** [comment]
yes, 2.10.x is intentional

**Comment by xeno-by** [comment]
Out of curiosity, how does this PR fare w.r.t the changes to typer introduced in master?

**Comment by lrytz** [comment]
what changes are you referring to?

**Comment by xeno-by** [comment]
Nothing in particular. When skimming through the diff, I immediately thought about modes, which were turned into value classes by Paul. Then I thought: "How many other conflicts there might be? Looks like a lot of work.".

**Comment by xeno-by** [comment]
After discussing this on the Scala meeting, we noticed that `AnnotationCheckers` has been changed in an incompatible way. @adriaanm @gkossakowski Can we afford this for 2.10.x?

**Comment by retronym** [comment]
Can you please [word wrap](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) your commit comments?

**Comment by lrytz** [comment]
right, merging 2.10.x into master will be painful.

**Comment by lrytz** [comment]
sure, will word wrap.

**Comment by gkossakowski** [comment]
@lrytz can you clarify why this needs to target 2.10.x and not master? Also, I'm not familiar with annotation checkers functionality to comment if incompatible changes can be merged. Is this api used by compiler plugins only?

**Comment by lrytz** [comment]
I'd like it to target 2.10 so that I can eventually release my compiler plugin for effects compatible with a 2.10 release compiler.

As for compatibility, yes it's only for compiler plugins. I'm pretty sure there is no compiler plugin out there except CPS that uses AnnotationCheckers (which I fixed already to use an AnalyzerPlugin). But if it's a concern, it should be possible to keep AnnotationCheckers source compatible I guess.

Edit: as discussed in the meeting, i'll try to come up with a source compatible patch. "people are doing everything you might not expect that anybody does"

File: src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+   *
+   * Note that the result is undefined if more than one annotation checker
+   * returns an adapted type which is not a subtype of `default`.
+   */
+  def pluginsTypedReturn(typer: Typer, tree: Return, pt: Type, default: Type): Type = {
+    val adaptedTypes = activePlugins flatMap { plugin =>
+      val adapted = plugin.pluginsTypedReturn(typer, tree, pt, default)
+      if (!(adapted <:< default)) List(adapted)
+      else List()
+    }
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:240]: Can you write conditions like this in the affirmative:

```
if (adapted <:< default) Nil else List(adapted)
```

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 2]
```diff
@@ -892,163 +902,257 @@ trait Namers extends MethodSynthesis {
...
+      // The info of both the module and the moduleClass symbols need to be assigned. monoTypeCompleter assigns
+      // the result of typeSig to the module symbol. The module class info is assigned here as a side-effect.
+      val result = templateSig(mdef.impl)
+      val pluginsTp = pluginsTypeSig(typer, mdef, WildcardType, result)
+      // Assign the moduleClass info (templateSig returns a ClassInfoType)
+      val clazz = moduleSym.moduleClass
+      clazz setInfo pluginsTp
+      // clazz.tpe returns a `ModuleTypeRef(clazz)`, a typeRef that links to the module class `clazz`
+      // (clazz.info would the ClassInfoType, which is not what should be assigned to the module symbol)
+      clazz.tpe
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:941]: Please prefer clazz.tpe_*, which is the same thing in this case but leaves us with one fewer call site which must someday be examined to glean the author's intent.

Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:941]: I would but we're on 2.10.x

File: src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala (hunk 1 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:240]: right, thanks - anyway i wanted to ask phaller why the `pluginsTypedReturn` has to be the way it is, i don't really get it.

**Comment by paulp** [comment]
I think the merge into master should be another pull request issued alongside the commit in cases like this one.

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala (hunk 2 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:941]: D'oh, I already can't keep them straight.

**Comment by lrytz** [comment]
ok, i can make one for master (the advantage being that you can exclude it from the 2.10 -> master merge, right?)

**Comment by paulp** [comment]
The advantage being that I don't have to do the merge when the change meet master. I don't actually think this one is that bad, but it would be a good idea for us to require it on general principles for nontrivial patches - leaving the merge for "unknown future party" potentially defers a huge amount of work, and merging isn't something which can be kicked down the road indefinitely. If it's hard at point X, it will only be harder at point X+1.

**Comment by lrytz** [comment]
@paulp should i do a pull request for master with all those commits, or only with the two main ones (Namers changes and Analyzer Plugins)? I'd prefer to include all if this doesn't complicate things later on.

**Comment by paulp** [comment]
If it's going into 2.10.x, then it's going into master eventually, so the most useful pull request merges the same commits into master which are merged into 2.10.x.

**Comment by lrytz** [comment]
ok, i'm happy to do that, but only after my commit gets merged into 2.10. as long as it's a pull request there can be rebases, and for every rebase i'd have to do the merge again from scratch.

**Comment by lrytz** [comment]
will make a new PR with the discussed changes.

**Pull request closed without merging**